jobmaster/dm/metadata/state.go

```diff
@@ -0,0 +1,129 @@
+package metadata
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/pingcap/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+
+	"github.com/hanfei1991/microcosm/pkg/metadata"
+)
+
+// State represents the state which need to be stored in metadata.
```

Comment:
Now we rely on `json.Marshal`, we should let developer knows the fields we want to store should be public and serializable (such as not a function, please take a look at when `json.Marshal` would fail). And I think it's also OK that we check it by reflection at `Put`